### ISSUE TITLE
Enable recursive load of multiple registry directories?

### ISSUE THREAD

**Issue by cixelsyd** [sparkleformation/sparkle_formation#93]
My project makes heavy use of "cfn-init" (unfortunately).  This means that I've got massive metadata sections for my machines, but it does mean that I can potentially re-use a bunch of the 'files', 'commands', etc from my metadata.

To facilitate that, I've been breaking the metadata down into bite sized registry entries ... but again, theres now a boat load of them... to the point where what I would like to do is organize them in a directory structure inside the registry folder.  I would like for SparkleFormation to load all .rb files inside of subdirectories of the main "registry" directory... but I don't see how to do that.  Otherwise, I'd like to specifically point at the registry files inside my subdirectories and have them loaded... but I don't see how to do that, either.

I've read through the docs multiple times now, but nothing there jumped out at me as the answer.

Thanks!
steve

**Comment by chrisroberts** [comment]
Hi! Allowing subdirectory use for organization _should_ be allowed and is simply a limitation of the current globbing pattern when loading up the files. #94 addresses this and it'll be available in the next release.

**Comment by cixelsyd** [comment]
You rock, Chris!  Thanks!

Steve


mobile

**Comment by cixelsyd** [comment]
Chris, the other thing that would be very helpful when we enable this
subdirectory loading scheme is to be very clear what the order of
precedence will be.

To be precise, I'm looking to load registry entries for my node "platform"
(windows or linux), and then load registry entries (of the same name) for
my node "category" (sql server, or apache, or Domain Controller) where I
have more specific or different instructions, and then last load node
"instance" on top of that for the rare cases where I care that node "ad01"
is the Primary Domain Controller.

This would let me move from the general to the most specific, while only
rewriting/restating certain things.  I would like to know the load order
for the directories so that I can ensure the general things are properly
subsumed by the more specific entries.




On Tue, Nov 3, 2015 at 12:58 PM, Steve Craig <cixelsydego@gmail.com> wrote:

> You rock, Chris!  Thanks!
>
> Steve
>
>
> mobile
>
> On Nov 2, 2015, at 5:21 PM, Chris Roberts <notifications@github.com>
> wrote:
>
> Hi! Allowing subdirectory use for organization *should* be allowed and is
> simply a limitation of the current globbing pattern when loading up the
> files. #94 <https://github.com/sparkleformation/sparkle_formation/pull/94>
> addresses this and it'll be available in the next release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sparkleformation/sparkle_formation/issues/93#issuecomment-153176264>
> .
>
>


-- 
http://linkedin.com/in/cixelsydego

(508) 372-0005 (cell)

**Comment by chrisroberts** [comment]
Hi @cixelsyd! This fix has been released as of v1.1.4. 

There is no defined load order of the files. It is a simple glob based iteration with no sorting applied. This is because SparklePacks (which is what the `sparkleformation` directory is) are loaded at initialization time but are not evaluated until requested. All items within a specific SparklePack must be unique (templates, dynamics, components, registry items) so if an item is defined twice (like two registry items with the same name) one will simply overwrite the other. There is a ticket open to kick back an error on this state #95 which will be included in the next release. 

What you are describing can be accomplished with the existing registry system. Instead of attempting to "merge" registry items as they are loaded, you can explicitly define the registry items and how they should behave. Registry items have full access to the DSL and all its helper methods, which means you can have registry items calling registry items and making explicit adjustments. For example:

```ruby
SfnRegistry.register!(:base_meta_init) do |*_|
  metadata('AWS::CloudFormation::Init') do
    _camel_keys_set(:auto_disable)
    config do
      commands('000_always_run') do
        command "DO_THINGS"
      end
    end
  end
end
```

We can assume that this command is something that gets run always, regardless of the destination. Now, if we wanted to add some linux stuff:

```ruby
SfnRegistry.register!(:linux_meta_init) do |*_|
  val = registry!(:base_meta_init)
  val.set!('AWS::CloudFormation::Init') do
    config.commands('100_linux_stuff') do
      command "LINUX_STUFF"
    end
  end
  val
end
```

So now we are building off the base item, and adding some linux "stuff". Taking a step further, we can add some app stuff (maybe apache):

```ruby
SfnRegistry.register!(:apache_linux_meta_init) do |*_|
  val = registry!(:linux_meta_init)
  val.set!('AWS::CloudFormation::Init') do
    config.commands('200_apache_stuff') do
      command "APACHE_STUFF"
    end
  end
  val
end
```

And now we end up with a "base" init registry item, a "linux" item that builds off the base providing some linuxy things, and an "apache" item that tosses in the apache bits at the end. This is just an example, and it's likely possible that this could be refactored better using a mix of dynamics and registry items, as well as features within cfn-init (like config sets). But from the more high level view of what you were asking about, this is one approach for breaking down and DRYing out registry items. With the fixed glob for loading, this means you can also isolate things into subdirectories for easier organization as well.

Hope that helps. If not just reply back here or you can catch me on freenode.

Cheers!
